core/src/main/java/org/apache/iceberg/MetricsUtil.java

```diff
@@ -456,4 +458,54 @@ public <T> void set(int pos, T value) {
       throw new UnsupportedOperationException("StructWithReadableMetrics is read only");
     }
   }
+
+  static class StructWithRefSnapshot implements StructLike {
+    private final StructLike struct;
+    private final Schema projectedSchema;
+    private final Snapshot snapshot;
+
+    private final int refSnapshotIdPosition;
+    private final int refSnapshotTimestampMillisPosition;
+
+    StructWithRefSnapshot(
+        StructLike struct,
+        Schema projectedSchema,
+        Snapshot snapshot,
+        Types.NestedField refSnapshotIdField,
+        Types.NestedField refSnapshotTimestampMillisField) {
+      this.struct = struct;
+      this.projectedSchema = projectedSchema;
+      this.snapshot = snapshot;
+      this.refSnapshotIdPosition =
+          Optional.ofNullable(refSnapshotIdField)
+              .map(f -> projectedSchema.columns().indexOf(refSnapshotIdField))
+              .orElse(-1);
+
+      this.refSnapshotTimestampMillisPosition =
+          Optional.ofNullable(refSnapshotIdField)
+              .map(f -> projectedSchema.columns().indexOf(refSnapshotTimestampMillisField))
+              .orElse(-1);
+    }
+
+    @Override
+    public int size() {
+      return projectedSchema.columns().size();
+    }
+
+    @Override
+    public <T> T get(int pos, Class<T> javaClass) {
```

Comment:
Wont this not work for Flink, the pos are not always at the end? See  https://github.com/apache/iceberg/pull/6222/files 

https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/MetricsUtil.java#L441

I think we should make a generic struct for this scenario now that we have so many fields.  ie, have a position map kind of like:  https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/util/StructProjection.java.  do you think its possible?